flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java

```diff
@@ -0,0 +1,313 @@
+				Assert.assertEquals("--port", ParameterProgram.actualArguments[2]);
+				Assert.assertEquals("1234", ParameterProgram.actualArguments[3]);
+
+				Assert.assertEquals(4, getExecutionConfig(jobGraph).getParallelism());
+
+				final SavepointRestoreSettings savepointRestoreSettings = jobGraph.getSavepointRestoreSettings();
+				Assert.assertTrue(savepointRestoreSettings.allowNonRestoredState());
+				Assert.assertEquals("/foo/bar", savepointRestoreSettings.getRestorePath());
+			}
+		);
+	}
+
+	@Test
+	public void testParameterPrioritization() throws Exception {
+		// configure submission via query parameters and JSON request, JSON should be prioritized
+		sendRequestAndValidateGraph(
+			handler,
+			restfulGateway,
+			() -> {
+				final JarRunRequestBody jsonRequest = new JarRunRequestBody(
+					ParameterProgram.class.getCanonicalName(),
+					"--host localhost --port 1234",
+					4,
+					true,
+					"/foo/bar"
+				);
+
+				final JarRunMessageParameters parameters = JarRunHeaders.getInstance().getUnresolvedMessageParameters();
+				parameters.allowNonRestoredStateQueryParameter.resolve(Collections.singletonList(false));
+				parameters.savepointPathQueryParameter.resolve(Collections.singletonList("/no/uh"));
+				parameters.entryClassQueryParameter.resolve(Collections.singletonList("please.dont.run.me"));
+				parameters.parallelismQueryParameter.resolve(Collections.singletonList(64));
+				parameters.programArgsQueryParameter.resolve(Collections.singletonList("--host wrong --port wrong"));
+
+				return createRequest(
+					jsonRequest,
+					parameters,
+					jarWithoutManifest
+				);
+			},
+			jobGraph -> {
+				Assert.assertEquals(4, ParameterProgram.actualArguments.length);
+				Assert.assertEquals("--host", ParameterProgram.actualArguments[0]);
+				Assert.assertEquals("localhost", ParameterProgram.actualArguments[1]);
+				Assert.assertEquals("--port", ParameterProgram.actualArguments[2]);
+				Assert.assertEquals("1234", ParameterProgram.actualArguments[3]);
+
+				Assert.assertEquals(4, getExecutionConfig(jobGraph).getParallelism());
+
+				final SavepointRestoreSettings savepointRestoreSettings = jobGraph.getSavepointRestoreSettings();
+				Assert.assertTrue(savepointRestoreSettings.allowNonRestoredState());
+				Assert.assertEquals("/foo/bar", savepointRestoreSettings.getRestorePath());
+			}
+		);
+	}
+
+	private static HandlerRequest<JarRunRequestBody, JarRunMessageParameters> createRequest(
+			JarRunRequestBody requestBody,
+			JarRunMessageParameters parameters,
+			Path jar) throws HandlerRequestException {
+
+		final Map<String, List<String>> queryParameterAsMap = parameters.getQueryParameters().stream()
+			.filter(MessageParameter::isResolved)
+			.collect(Collectors.toMap(
+				MessageParameter::getKey,
+				JarRunHandlerParameterTest::getValuesAsString
+			));
+
+		return new HandlerRequest<>(
+			requestBody,
+			JarRunHeaders.getInstance().getUnresolvedMessageParameters(),
```

Comment:
`HandlerRequest` doesn't work like that.
It takes the parameter map from netty and inserts them into the `MessageParameters`.
We could move logic into a static factory method and have the request be a simple container, but changing that would be out-of-scope of this PR.

Reply:
Ah I see. You're right. I think this could be a nice improvement to initialize the `MessageParameter` instance before giving it to the `HandlerRequest`. But this is out of scope for this PR.